server/mqtt.go

```diff
@@ -3835,6 +4262,86 @@ func (sess *mqttSession) cleanupFailedSub(c *client, sub *subscription, cc *Cons
 	}
 }
 
+func (sess *mqttSession) countQOS2Subs() int {
+	c := 0
+	for _, qos := range sess.subs {
+		if qos == 2 {
+			c++
+		}
+	}
+	return c
+}
+
+func (sess *mqttSession) processPubRelConsumer(c *client) (*ConsumerConfig, error) {
+	var ccDelete *ConsumerConfig
+	var cc *ConsumerConfig
+	create := false
+
+	sess.mu.Lock()
+	cQOS2 := sess.countQOS2Subs()
+	if cQOS2 == 0 && sess.pubrelConsumer != nil {
+		ccDelete, sess.pubrelConsumer = sess.pubrelConsumer, nil
+	}
+	if cQOS2 > 0 && sess.pubrelConsumer == nil {
+		create = true
+	}
+	cc = sess.pubrelConsumer
+	sess.mu.Unlock()
+
+	deliverSubj := sess.pubRelDeliverySubject()
+	deliverSubjB := []byte(deliverSubj)
+	if ccDelete != nil {
+		sess.deleteConsumer(ccDelete)
```

Comment:
check the order, session clear

Reply:
split `process` into a create and delete (only in clear session).

Reply:
Agreed